### PR TITLE
nixos/nsswitch: Make databases more configurable

### PR DESCRIPTION
**Pull request by dasJ**
###### Motivation for this change

Instead of hardcoding all nss modules that are added into nsswitch,
there are now options exposed.
This allows users to add own nss modules (I had this issue with
winbindd, for example).
Also, nss modules could be moved to their NixOS modules which would
make the nsswitch module slimmer.

As the lists are now handled by the modules system, we can use mkOrder
to ensure a proper order as well as mkForce to override one specific
database type instead of the entire file.

cc @Infinisil to make sure I use `mk*` correctly
cc @LnL7 because you were the last to touch the file and I drop your option

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/config/nsswitch.nix [hunk 1]
```diff
@@ -10,35 +10,44 @@ let
   canLoadExternalModules = config.services.nscd.enable;
   myhostname = canLoadExternalModules;
   mymachines = canLoadExternalModules;
+  # XXX Move these to their respective modules
```
Comment (flokli) [inline nixos/modules/config/nsswitch.nix:13]: Could these also be moved into the individual systems.

We have a similar mess inside the pam module, and it'd be nice to clean all this up.

Comment (dasJ) [reply nixos/modules/config/nsswitch.nix:13]: Yeah, that's what I wanted to say

**Comment by dasJ** [comment]
Ah example from my Samba DC config:
```nix
{
    system.nssDatabases = rec {
      passwd = [ "winbind" ];
      group = passwd;
    };
}
```

Results in

```
passwd:    files winbind mymachines systemd
group:     files winbind mymachines systemd
…
```

File: nixos/modules/config/nsswitch.nix [hunk 2]
```diff
@@ -10,35 +10,44 @@ let
...
-
-  shadowArray = [ "files" ]
-    ++ optional sssd "sss"
-    ++ optional ldap "ldap";
-
-  servicesArray = [ "files" ]
-    ++ optional sssd "sss";
+  hostArray = mkMerge [
+    (mkBefore [ "files" ])
+    (optional mymachines "mymachines")
```
Comment (Infinisil) [inline nixos/modules/config/nsswitch.nix:23]: Prefer `mkIf` over `optional` in NixOS modules

**Review by infinisil — CHANGES_REQUESTED** [review]

File: nixos/modules/config/nsswitch.nix (hunk 2 above)
Comment (dasJ) [reply nixos/modules/config/nsswitch.nix]: Thanks, should be fixed

File: nixos/modules/config/nsswitch.nix (hunk 1 above)
Comment (flokli) [reply nixos/modules/config/nsswitch.nix:13]: Can you add a commit on top that does this, for some of the NSS modules?

I'd assume the systemd and oslogin nss modules should be a somewhat trivial example, and then we could create a follow-up ticket tracking the migration of the others.

Comment (dasJ) [reply nixos/modules/config/nsswitch.nix:13]: Just added the commit. I used sss instead, because it allows me to drop the `servicesArray` ;)

**Comment by flokli** [comment]
Thanks!

**Pull request merged**